lib/std/hash/wyhash.zig

```diff
@@ -1,209 +1,207 @@
+    inline fn read(comptime bytes: usize, data: []const u8) u64 {
+        std.debug.assert(bytes <= 8);
+        const T = std.meta.Int(.unsigned, 8 * bytes);
+        return @as(u64, std.mem.readIntLittle(T, data[0..bytes]));
     }
-};
-
-/// Fast non-cryptographic 64bit hash function.
-/// See https://github.com/wangyi-fudan/wyhash
-pub const Wyhash = struct {
-    state: WyhashStateless,
 
-    buf: [32]u8,
-    buf_len: usize,
+    inline fn mum(a: *u64, b: *u64) void {
+        const x = @as(u128, a.*) *% b.*;
+        a.* = @truncate(u64, x);
+        b.* = @truncate(u64, x >> 64);
+    }
 
-    pub fn init(seed: u64) Wyhash {
-        return Wyhash{
-            .state = WyhashStateless.init(seed),
-            .buf = undefined,
-            .buf_len = 0,
-        };
+    inline fn mix(a_: u64, b_: u64) u64 {
+        var a = a_;
+        var b = b_;
+        mum(&a, &b);
+        return a ^ b;
     }
 
-    pub fn update(self: *Wyhash, b: []const u8) void {
-        var off: usize = 0;
+    // Input must reside in a 16-byte buffer. The input slice passed be offset into it in which
+    // case this function will index in front of the slice.
+    inline fn final1(self: *Wyhash, input: []const u8) void {
+        std.debug.assert(input.len < 48);
+        self.state[0] ^= self.state[1] ^ self.state[2];
 
-        if (self.buf_len != 0 and self.buf_len + b.len >= 32) {
-            off += 32 - self.buf_len;
-            @memcpy(self.buf[self.buf_len..][0..off], b[0..off]);
-            self.state.update(self.buf[0..]);
-            self.buf_len = 0;
+        var i: usize = 0;
+        while (i + 16 < input.len) : (i += 16) {
+            self.state[0] = mix(read(8, input[i..]) ^ secret[1], read(8, input[i + 8 ..]) ^ self.state[0]);
         }
 
-        const remain_len = b.len - off;
-        const aligned_len = remain_len - (remain_len % 32);
-        self.state.update(b[off .. off + aligned_len]);
-
-        const src = b[off + aligned_len ..];
-        @memcpy(self.buf[self.buf_len..][0..src.len], src);
-        self.buf_len += @intCast(u8, b[off + aligned_len ..].len);
+        // Possible lookbehind past pointer start.
+        self.a = read(8, (input.ptr + input.len - 16)[0..8]);
+        self.b = read(8, (input.ptr + input.len - 8)[0..8]);
     }
 
-    pub fn final(self: *Wyhash) u64 {
-        const rem_key = self.buf[0..self.buf_len];
-
-        return self.state.final(rem_key);
+    inline fn final2(self: *Wyhash) u64 {
+        self.a ^= secret[1];
+        self.b ^= self.state[0];
+        mum(&self.a, &self.b);
+        return mix(self.a ^ secret[0] ^ self.total_len, self.b ^ secret[1]);
     }
 
     pub fn hash(seed: u64, input: []const u8) u64 {
-        return WyhashStateless.hash(seed, input);
+        var self = Wyhash.init(seed);
+
+        if (input.len <= 16) {
+            self.smallKey(input);
+        } else {
+            var i: usize = 0;
+            while (i + 48 <= input.len) : (i += 48) {
```

Comment:
Incorrect for block sizes of 48. This should be `i + 48 < input.len`. The last 48-bytes is always run through `final1`. This is trivial here but is somewhat annoying in the iterative api. All other hash functions process blocks as they are filled, wyhash however is conditional in the sense that you need to hold off processing blocks if you are aligned with them internally in case the user calls final and the standard round is not run, but the final1 function instead.